internal/mongodoc/resource.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongodoc
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charm.v6-unstable/resource"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// NewResourceQuery formats the provided details into a mongo query for
```

Comment:
Other than a very few things, we've tried to keep business logic outside of the mongodoc package. The place for this is in the internal/charmstore package (and even then it might not warrant exporting, depending on what requirements are).